docs/book/content/types/objects/error_handling.md

```diff
@@ -169,3 +178,293 @@ The specified structured error information is included in the [`extensions`](htt
+
+### Example Input Validation (simple)
+
+In this example, a basic input validation is implemented with GraphQL
+types. Strings are used to identify the problematic field name. Errors
+for a particular field are also returned as a string. In this example,
+the string contains a localized error message. However, it is also
+possible to return a unique string identifier.
+
+```rust
+#[derive(juniper::GraphQLObject)]
+pub struct Item {
+    name: String,
+    quantity: i32,
+}
+
+#[derive(juniper::GraphQLObject)]
+pub struct ValidationError {
+    field: String,
+    message: String,
+}
+
+#[derive(juniper::GraphQLObject)]
+pub struct ValidationErrors {
+    errors: Vec<ValidationError>,
+}
+
+#[derive(juniper::GraphQLUnion)]
+pub enum GraphQLResult {
+    Ok(Item),
+    Err(ValidationErrors),
+}
+
+pub struct Mutation;
+
+#[juniper::graphql_object]
+impl Mutation {
+    fn addItem(&self, name: String, quantity: i32) -> GraphQLResult {
+        let mut errors = Vec::new();
+
+        if !(10 <= name.len() && name.len() <= 100) {
+            errors.push(ValidationError {
+                field: "name".to_string(),
+                message: "between 10 and 100".to_string()
+            });
+        }
+
+        if !(1 <= quantity && quantity <= 10) {
+            errors.push(ValidationError {
+                field: "quantity".to_string(),
+                message: "between 1 and 10".to_string()
+            });
+        }
+
+        if errors.is_empty() {
+            GraphQLResult::Ok(Item { name, quantity })
+        } else {
+            GraphQLResult::Err(errors)
+        }
+    }
+}
+
+```
+
+Each function may have a different return type. Depending on the input
+parameters a new result type is required. For example, adding a user
+require a new result type which contains the variant `Ok(User)`
+instead of `Ok(Item)`.
+
+Finally it is possible to send a mutation request and handle the
+result. The following example query shows how to handle the result.
+
+```graphql
+{
+  mutation {
+    addItem(name: "", quantity: 0) {
+      ... on Item {
+        name
+      }
+      ... on ValidationErrors {
+        errors {
+          field
+          message
+        }
+      }
+    }
+  }
+}
+```
+
+A useful side effect of this approach is to have partially successful
+paths. Therefore, a if one of multiple paths fails, the result of the
+successful paths are not discarded.
+
+### Example Input Validation (complex)
+
+Instead of using strings do propagated error, it is possible to use
+GraphQL's type system to describe the errors more precisely. For each
+failable input variable a field in a GraphQL object is created. The
+field is set if the validation for that particular field fails. Notice
+that some kind of code generation reduces the unnecessary work. The
+amount of types which are required is significant larger than
+before. Each functions has their custom `ValidationResult` which
+contains only fields provided by the function.
```

Comment:
```suggestion
Instead of using strings to propagate errors, it is possible to use
GraphQL's type system to describe the errors more precisely.

For each fallible input variable a field in a GraphQL object is created. The
field is set if the validation for that particular field fails. You will likely want some kind of code generation to reduce repetition as the number of types required is significantly larger than
before. Each resolver function has a custom `ValidationResult` which
contains only fields provided by the function.
```